# Prosjektoppgave del 2 - Statistisk kvalitetskontroll
Mål med oppgaven:
- Deskripttiv statistikk
- Tolkning av kapabilitetsindeks
- Lage Shewart $\overline{x}$- og $s$-diagram.
- Bruke og tolke Shewart $\overline{x}$- og $s$-diagram.


En bedrift er ansvarlig for sløying og pakking av fisk. 

De ønsker å foreta kontrollmålinger for å sikre at vekten av hver fiskefilet  et ferdigpakket produkt er korrekt.<br>Anta at bedriften hver dag gjennomfører en stikkprøve der de veier $n=8$ tilfeldige valgte fiskefileter.  

##  Oppgave 1, oppstartsfasen.
Vi er nå i den fasen at prosessen forhåpentligvis er klar til oppstart - For eksempel er forsøksplanlegging en del av forarbeidet for å få riktige innstillinger.

Vi simulerer at bedriften kjører en testrunde med stikkprøver der $n=8$ tilfeldige valgte fileeter veies. Bedriften gjør $k=12$ slike stikkprøver.

Resultatene er oppgitt i  tabellen du produserer under der hver rad svarer til en stikkprøve. Vekten er oppgitt i gram.

For at hver gruppe skal få sitt eget datasett legger du først inn nummeret på gruppa, dette som skal være et frø i simuleringa

In [8]:
Gruppenr = xx # Her skal xx erstattes med nummeret på gruppa di

n = 8
k = 12
tot = n*k

mu0 = 125.20  #Vi simulerer at de  har klart å innstille ganske godt men ikke helt nøyaktig på 125 gram
sigma0 = 1.40

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

#pip install tabulate
np.random.seed(seed = Gruppenr)

testdata = pd.DataFrame(mu0+sigma0*np.random.randn(k,n))
testdata=testdata.round(1)

from IPython.display import HTML, display
display(testdata)    #Viser alle dataene

sammendrag=testdata.apply(pd.DataFrame.describe,axis=1)
display(sammendrag)  # Et sammendrag av forventningsverdi, standardavvik mm. for hver stikkprøve








0      1      2      3      4      5      6      7
0   124.6  124.7  125.3  124.1  124.9  124.1  124.1  127.8
1   124.2  125.1  125.6  125.0  123.8  123.9  126.9  124.7
2   122.7  125.3  124.4  124.2  123.9  127.3  128.0  124.6
3   125.0  124.9  124.1  125.3  126.1  124.8  125.9  126.2
4   126.7  125.1  124.6  124.9  125.2  123.5  123.7  123.7
5   125.2  125.1  124.3  127.4  124.6  124.5  124.2  127.0
6   127.2  124.3  125.7  126.0  123.2  122.6  123.0  128.6
7   125.5  125.6  124.9  124.5  127.7  123.9  125.6  128.4
8   124.7  125.0  124.7  123.5  124.9  127.0  124.7  125.2
9   123.4  121.8  126.6  123.1  124.4  124.6  127.2  123.3
10  126.3  122.3  126.2  126.7  125.5  125.6  125.4  126.4
11  124.2  128.8  124.4  124.9  124.8  125.0  124.0  124.5

count      mean       std    min      25%     50%      75%    max
0     8.0  124.9500  1.230563  124.1  124.100  124.65  125.000  127.8
1     8.0  124.9000  1.019804  123.8  124.125  124.85  125.225  126.9
2     8.0  125.0500  1.773616  122.7  124.125  124.50  125.800  128.0
3     8.0  125.2875  0.731803  124.1  124.875  125.15  125.950  126.2
4     8.0  124.6750  1.062006  123.5  123.700  124.75  125.125  126.7
5     8.0  125.2875  1.235704  124.2  124.450  124.85  125.650  127.4
6     8.0  125.0750  2.162505  122.6  123.150  125.00  126.300  128.6
7     8.0  125.7625  1.541741  123.9  124.800  125.55  126.125  128.4
8     8.0  124.9625  0.967969  123.5  124.700  124.80  125.050  127.0
9     8.0  124.3000  1.824437  121.8  123.250  123.90  125.100  127.2
10    8.0  125.5500  1.394889  122.3  125.475  125.90  126.325  126.7
11    8.0  125.0750  1.544345  124.0  124.350  124.65  124.925  128.8

### Oppgave 1 a)
Visualiser dataene (betraktet som et datasett med $8\cdot12 = 96$ enkeltdata) som histogram og med dataene i rekkefølge. <br>
Diskuter kort hva du ser.

In [6]:
# Utregning og plott til oppgave 1a)
# Konverterer verdiene i tabellen til en vektor:
data_som_vektor = testdata.to_numpy().flatten()
# Fortsett selv

Hva ser du fra plottene?

### Oppgave 1b)
Regn ut gjennomsnitt $\overline{x}$  og empirisk standardavvik $s$ for HELE datasettet samlet.

De "teoretisk riktige verdiene"  $\mu$ og $\sigma$ er brukt  i simuleringen av dataene, men disse er egentlig ukjente og skal ikke brukes i besvarelsen. 

Dermed er $\overline{x}$  og empirisk standardavvik $s$ vårt beste forslag til hva gjennomsnitt og standardavvik er i det lange løp (om prosessen ikke endrer seg).

In [ ]:
# Svar på 1b)


### Oppgave 1c)

Regn ut gjennomsnitt og standardavvik i hvert enkelt av de $k$ forsøkene.

Regn ut gjennomsnitt og standardavvik i hvert enkelt av de $k$ forsøkene.


Regn ut $\overline{\overline{x}}$, gjennomsnitt av gjennomsnittene i enkeltstikkprøvene.<br> Er dette det samme som gjennomsnittet av alle dataene (fra b-opppgaven)?

Regn ut $\overline{s^2}$, gjennomsnittet av variansene  for de $k=12$ stikkprøvene, og standardavviket på basis av dette. <br> Er dette det samme som å regne ut variansen og standardavviket for hele datasettet?



In [9]:
# Utregning (og tallsvar)  på  1c)
xbar=testdata.mean(axis=1) 
s=testdata.std(axis=1)
# Fortsett selv




Tolkning av resultatet i 1c

### Oppgave 1d)
En storkunde  har stilt krav om toleransegrense $T_L=120$ gram og $T_u=130$ gram.

Beregn kapabilitet og kapabilitetsindeks.<br>
Tolk kapabilitetsindeksen.

(Om resultatet blir at kundens krav ikke  er oppfylt skal du tenke deg at vi forhandler fram en løsning med litt redusert pris, slik at dere går videre med oppgaven med forutsetningene fra oppgave 1 uansett).




In [ ]:
# Utregninger til 1d)

Tolkning til oppgave 1d)

## Oppgave 2,  Kvalitetskontroll med Shewart $\overline{x}$-diagram
Vi tenker oss nå at oppstartsfasen fra oppgave 1 er unnagjort, og at innstillingene fra denne er bra nok.

Vi skal da først lage Shewart-diagrammer, og så bruke disse i en ordinær produksjonsfase. <br>
Vi skal  holde rede på om innstillingene fortsatt er i orden, ved hjelp av daglige stikkprøver på vekten av $n=8$ fileter. 


### Oppgave 2a
Bedriften ønsker å kontrollere om det inntreffer kvalitetsavvik ved å se på gjennomsnittsvekten av $n=8$ fileeter i en stikkprøve. <br>Formuler situasjonen som en hypotese (som skal testes i $\overline{x}$-diagram).

Forutsett i denne deloppgaven at gjennomsnittvekten etter en stund har avtatt til 124 gram (uendret standardavvik).  

Hva er da forventet antall dager (stikkprøver) før første registrerte kvalitetsavvik inntreffer ved bruk av Shewart $\overline{x}$-diagram? 


Oppgave 2a, formulering av hypotesetesten:

In [10]:
# Oppgave 2a, utregning av forventet antall dager (stikkprøver) før første feilaktige stopp av produksjonen:

### Oppgave 2b
Regn ut øvre og nedre kontrollgrenser (OKG og NKG) og øvre og nedre varselgrenser (OVG og NVG) i Shewart $\overline{x}$-diagram basert på dataene i oppgave 1.

In [ ]:
# Løsning oppg 2b

### Simulering av tenkt forløp av produksjonen til resten av oppgaven:

In [11]:
# Det er ikke nøsvendig å forstå hvordan disse dataene er lagd.
# Tvert i mot er det meningen du skal tro jeg har vært ute og veid 320 fisker.

n=8        # Fortsatt 8 veiinger i hver stikkprøve
k_ny= 40   # Skjema skal lages for 40 stikkprøver (dager)

np.random.seed(seed = Gruppenr+1)
A= np.zeros((k_ny,n))
for i in range(0,k_ny):
   for j in range(0,n):
    A[i,j]= round(125.3 - i**2/1000 + 1.2*(1+i/20)*np.random.randn(),1)
    
Stikkprovedata=pd.DataFrame(A)

sammendrag=Stikkprovedata.apply(pd.DataFrame.describe,axis=1)
display(sammendrag)

### Oppgave 2c
Dataene i "Stikkprovedata" simulert over er resultatene av en tenkt stikkprøveserie (med en stikkprøve per dag) i 40 dager, bortsett fra at i virkeligheten vil testserien stoppes og overhaling eller justering av produksjonsprosessen gjennomføres om Shewart-diagrammene tilsier dette. 

Lag først et enkelt Shewart $\overline{x}$-diagram der du har med alle 40 gjennomsnittene (og dermed ignorerer at prosessen burde vært stoppet). Bruk diagrammet til å anslå ved hvilken stikkprøve (nummerert fra 0)  prosessen skulle bli stoppet i følge $\overline{x}$-diagrammet?

In [ ]:
# Utregninger til 2b)

Konklusjon, oppgave 2c:
    

### Oppgave 2d
Lag en litt mer forseggjort  figur med  Shewart $\overline{x}$-diagram med kontroll- og varselsgrenser kargjort for $k=40$ stikkprøver, men dataene er bare fylt ut til og med den dagen prosessen ville stoppet om dette var en virkelig kvalitetskontroll.

In [ ]:
# Utregning/plot til oppgave 2c

## Oppgave 3, $s$-diagram for spredning.
I denne oppgaven skal dere gjøre omtrent det samme som for $\overline{x}$-diagrammet, bortsett fra at det er et spredningdiagram som skal brukes.

### Oppgave 3a
Regn ut øvre og nedre kontrollgrense i $s$-diagram fra forsøksdataene i oppgave 1.


In [ ]:
# Utregninger til oppgave 3a

### Oppgave 3b
Lag et enkelt spredingsdiagram ($s$-diagram) med kontrollgrensene og alle standardavvikene for de 40 stikkprøvene.

Bruk dette til å anslå hvilken dag prosessen egentlig burde vært stoppet i følge $s$-diagrammet.

In [ ]:
# Utregning/plot til oppgave 3b

### Oppgave 3c
Lag en littt finere figur med  $s$-diagram kargjort for $k=40$ stikkprøver, men dataene bare fylt ut til og med den dagen (stikkprøven) gir at produksjonen må stoppes for rejustering. 

Hvilken dag burde prosessen vært stoppet om vi brukte både $\overline{x}$ og $s$-diagram (som man vanligvis gjør).

In [ ]:
# Utregning/plot til oppgave 3c

Oppgavesettet er laget av Ketil Arnesen i 2020 og revidert av Hans Petter Hornæs i 2021.